discovery: pinned syncers

cfromknecht opened the pull request:
A pinned syncer is an ActiveSyncer that is configured to always remain
active for the lifetime of the connection. Pinned syncers do not count
towards the total NumActiveSyncer count, which are rotated periodically.

This features allows nodes to more tightly synchronize their routing
tables by ensuring they are always receiving gossip from distinguished
subset of peers.

wpaulino commented on config.go:
```diff
@@ -309,6 +309,8 @@ type Config struct {
 
 	Routing *lncfg.Routing `group:"routing" namespace:"routing"`
 
+	Gossip *lncfg.Gossip `group:"gossip" namespace:"gossip"`
```
Thoughts on moving the existing `numgraphsyncpeers` flag here and deprecating the top-level one while still parsing it? Then we can remove the top-level one at a later release.

cfromknecht replied:
sgtm, let's do that in a follow up

wpaulino commented on lntest/itest/lnd_test.go:
```diff
@@ -10335,16 +10335,39 @@ func testGraphTopologyNotifications(net *lntest.NetworkHarness, t *harnessTest)
+	err = net.EnsureConnected(ctxt, alice, bob)
+	require.NoError(t.t, err)
+
+	// Alice stimmy.
```
:joy:

cfromknecht commented on discovery/sync_manager.go:
```diff
@@ -140,6 +145,10 @@ type SyncManager struct {
 	// currently receiving new graph updates from.
 	inactiveSyncers map[route.Vertex]*GossipSyncer
 
+	pinnedSyncers map[route.Vertex]struct{}
```
removed

cfromknecht commented on discovery/sync_manager.go:
```diff
@@ -325,7 +326,9 @@ func (m *SyncManager) syncerHandler() {
 			// when the historical sync completed.
-			setHistoricalSyncer(s)
+			if !isPinnedSyncer {
+				setHistoricalSyncer(s)
```
i tried going down that path but it gets messy, needing to select/transition between pinned vs unpinned peers to be the next historical syncer and test all the edge cases. it was much simpler just to treat them as separate.

cfromknecht commented on discovery/sync_manager.go:
```diff
@@ -386,13 +388,15 @@ func (m *SyncManager) syncerHandler() {
 			// If we've already performed an initial historical
 			// sync, we won't try again with our passive peers.
 			if m.IsGraphSynced() {
+				m.cfg.HistoricalSyncTicker.Pause()
```
removed the pauses in this area. the ticker is now only reset when setting the initial syncer. after that completes it remains active permanently.

cfromknecht commented on discovery/sync_manager.go:
```diff
@@ -380,12 +380,16 @@ func (m *SyncManager) syncerHandler() {
-			s := m.forceHistoricalSync()
+			// If we've already performed an initial historical
+			// sync, we won't try again with our passive peers.
+			if m.IsGraphSynced() {
```
Gotcha that makes sense, reversed the ordering

cfromknecht commented on lnrpc/rpc.proto:
```diff
@@ -1487,6 +1487,13 @@ message Peer {
+        */
+        PINNED_ACTIVE_SYNC = 3;
+
+        PINNED_PASSIVE_SYNC = 4;
```
good call, ended up removing and now there's only `PINNED_SYNC`. we no longer need the two states with the introduction of `syncerIdle`

wpaulino reviewed (approved):
LGTM. Looks like there's an flake (channel not opened before timeout) in one of the integration tests. I'm assuming it's the pre-existing one which @halseth is working on fixing with https://github.com/lightningnetwork/lnd/pull/4953, though it'd be nice to double check.

halseth commented on discovery/sync_manager.go:
```diff
@@ -214,9 +213,12 @@ func (m *SyncManager) syncerHandler() {
 		initialHistoricalSyncer = s
 		initialHistoricalSyncSignal = s.ResetSyncedSignal()
+
+		m.cfg.HistoricalSyncTicker.Pause()
```
add a comment to why this is done?

halseth commented on discovery/sync_manager.go:
```diff
@@ -322,7 +323,9 @@ func (m *SyncManager) syncerHandler() {
 			// handle disconnects. We'll also use a signal to know
 			// when the historical sync completed.
-			setInitialHistoricalSyncer(s)
+			if !isPinnedSyncer {
```
hm, why this condition?

cfromknecht replied:
Several reason, also mentioned in another comment w/ wilmer. 

Each pinned peer already does a historical sync, so there can be many of them at the same time. When it comes to replacement, ideally you'd prioritize pinned over passive syncers but the bookkeeping gets tricky when they're all happening in parallel. Disconnects need to be handled differently depending on pinned vs regular. Tl;dr the number of edge cases isn't worth it, and it's simpler to reason about them as separate peer groups.

My preferred approach if we are interested in the cosmetic aspect is just a callback that syncers can use to report they've completed a historical sync.

halseth replied:
Yeah, I'm kinda worried that all these different states can become hard to reason about and maintain. Perhaps not in the scope of this PR, but it could use an overhaul

halseth commented on lntest/itest/lnd_test.go:
```diff
@@ -10330,19 +10330,77 @@ func subscribeGraphNotifications(t *harnessTest, ctxb context.Context,
+	require.NoError(t.t, err)
+	defer shutdownAndAssert(net, t, alice)
+
+	// When testing pinnned syncing, we need to ensure Alice doesn't select
```
I don't get this. If Bob is the pinned peer, why do we need to check that Bob won't also be the active peer? Aren't we checking that Alice is syncing from dummy below, while we should be checking she syncs from Bob?

cfromknecht replied:
Alice will complete an initial sync with the dummy, which means her graph will also be empty (but `synced_to_graph` will show true).

This is mainly done to avoid false positives due to regressions, since the test will pass all the same if we are unintentionally in ACTIVE_SYNC with Bob.

cfromknecht replied:
opted for more explicit check of the sync state, ptal

halseth commented on discovery/sync_manager.go:
```diff
@@ -140,6 +145,8 @@ type SyncManager struct {
 	// currently receiving new graph updates from.
 	inactiveSyncers map[route.Vertex]*GossipSyncer
 
+	pinnedActiveSyncers map[route.Vertex]*GossipSyncer
```
add a doc similar to other maps

halseth commented on lncfg/gossip.go:
```diff
@@ -0,0 +1,30 @@
+		if err != nil {
+			return err
+		}
+		fmt.Printf("vertex: %v\n", vertex)
```
remove?

halseth reviewed (approved):
LGTM. Not sure if this should go into v0.12.1 though, as it is new functionality.

cfromknecht commented:
addressed comments, also moved `markGraphSynced` into the syncer so that cosmetically we show `synced_to_graph` whenever any peer (pinned or regular) finishes a historical sync

cfromknecht commented:
> Not sure if this should go into v0.12.1 though, as it is new functionality.

We have released new features in point releases before, main thing is that this change is not breaking since it is opt-in.

In this case, users have reported an inability to synchronize the routing tables of two directly-connected nodes. This occurs because when the number of open channels is high because the historical sync randomization takes a long time before selecting their node again.
